tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -2936,6 +2941,103 @@ def unique(self):
 
     return _UniqueDataset(self)
 
+  def rejection_resample(self,
+                         class_func,
+                         target_dist,
+                         initial_dist=None,
+                         seed=None):
+    """A transformation that resamples a dataset to achieve a target distribution.
+
+    Lets consider the following example where a dataset with an initial data
+    distribution of `init_dist` needs to be resampled into a dataset with
+    `target_dist` distribution.
+
+    >>> import collections
+    >>> init_dist = [0.5, 0.5]
+    >>> target_dist = [0.6, 0.4]
+    >>> num_classes = len(init_dist)
+    >>> num_samples = 100000
+    >>> data_np = np.random.choice(num_classes, num_samples, p=init_dist)
+    >>> dataset = tf.data.Dataset.from_tensor_slices(data_np)
+    >>> x = collections.defaultdict(int)
+    >>> for i in dataset:
+    ...   x[i.numpy()] += 1
+
+    The value of `x` will be close to `{0: 50000, 1: 50000}` as per the
+    `init_dist` distribution.
+
+    >>> dataset = dataset.rejection_resample(
+    ...    class_func=lambda x: x % 2,
+    ...    target_dist=target_dist,
+    ...    initial_dist=init_dist)
+
+    >>> y = collections.defaultdict(int)
+    >>> for i in dataset:
+    ...   cls, _ = i
+    ...   y[cls.numpy()] += 1
+
+    The value of `y` will be now be close to `{0: 75000, 1: 50000}` thus
+    satisfying the `target_dist` distribution.
+
+    Args:
+      class_func: A function mapping an element of the input dataset to a scalar
+        `tf.int32` tensor. Values should be in `[0, num_classes)`.
+      target_dist: A floating point type tensor, shaped `[num_classes]`.
+      initial_dist: (Optional.)  A floating point type tensor, shaped
+        `[num_classes]`.  If not provided, the true class distribution is
+        estimated live in a streaming fashion.
+      seed: (Optional.) Python integer seed for the resampler.
+
+    Returns:
+      A `Dataset`
+    """
+
+    target_dist_t = ops.convert_to_tensor(target_dist, name="target_dist")
```

Comment:
Moving all of our implementations into `dataset_ops.py` is making `dataset_ops.py` quite long and hard to navigate. In later PRs we should consider moving the dataset transformations from `dataset_ops.py` into their own files, similar to what we do for experimental ops. It would also make graduating experimental ops more straightforward. @jsimsa what are your thoughts?

Reply:
This is a valid concern and I think it would be a useful refactor. That said, we should do this in a manner which does not require updating callsites that import `dataset_ops`. One option would be that we keep `dataset_ops.py` as a shim to import symbols from the "per transformation" modules.

Reply:
@aaudiber @jsimsa we might also want to consider the circular dependencies and file duplications during the promotion/refactor process if we are going to maintain separate files per transformation and use them in `dataset_ops.py`. Also, if we are using `dataset_ops.py` as a shim where the API layout is unaffected for the users, we might have to maintain a new file where the circular dependencies can be prevented. WDYT?

Reply:
+1 for making `dataset_ops` a shim. The refactor should have no user-facing impact. We can use LazyLoader in the dataset impl files to handle the circular reference on `dataset_ops.py`. Keeping `dataset_ops` readable is much more important than avoiding circular dependencies between dataset implementations and `dataset_ops`.

Reply:
okay sounds good!